ast/impl/pretty_printer.cpp

```diff
@@ -0,0 +1,320 @@
+#include "fmt/core.h"
+#include "visitors.hpp"
+#include <memory>
+
+using namespace ast;
+
+extern std::map<lexer::TokenType, std::string> g_opToString;
+
+namespace san {
+
+void PrettyPrinter::visit(Program* node) {
+    if (node == nullptr) {
+        return;
+    }
+
+    for (auto type : node->types) {
+        type->accept(*this);
+        newline();
+        newline();
+    }
+
+    for (auto variableDecl : node->variables) {
+        variableDecl->accept(*this);
+        newline();
+    }
+
+    newline();
+
+    for (auto routine : node->routines) {
+        routine->accept(*this);
+        newline();
+        newline();
+    }
+}
+
+void PrettyPrinter::visit(RoutineDecl* node) {
+    if (node == nullptr) {
+        return;
+    }
+
+    fmt::print("routine {} (", node->name);
+
+    m_oneLine++;
+
+    for (size_t i = 0; i < node->parameters.size(); i++) {
+        node->parameters[i]->accept(*this);
```

Comment:
This may not work out so nicely since the type of `RoutineDecl::parameters` changed from `vector<Parameter>` to `vector<VariableDecl>`, which will cause it to print the keyword `var` in the parameters